### ISSUE TITLE
[cfg lowering] missing basic block in irreducible flow graph

### ISSUE THREAD

**Issue by oroppas** [flang-compiler/f18-llvm-project#870]
`bbc` hard crashed.

```fortran
      SUBROUTINE CMAUTO

      INTEGER IBITS(2)
      COMMON /SYSTEM/ NLPP

      NLIN  = 1000

      IF (NLIN .LT. NLPP) GO TO 220
  200 NLIN = 0
  220 CALL BITPAT(CE(1),IBITS)
      IF (NLIN .GT. NLPP) GO TO 200

      END
```

Here's the stack dump

```
Stack dump:
0.	Program arguments: /home/ryuta/packages/flang/build/bin/bbc ./crash.f
 #0 0x00007fcec0eafac1 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/ryuta/packages/flang/build/bin/../lib/libLLVMSupport.so.13git+0x1e2ac1)
 #1 0x00007fcec0ead86e llvm::sys::RunSignalHandlers() (/home/ryuta/packages/flang/build/bin/../lib/libLLVMSupport.so.13git+0x1e086e)
 #2 0x00007fcec0eb01f6 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007fcec4ebd870 __restore_rt sigaction.c:0:0
 #4 0x00007fcec30338d5 mlir::Operation::create(mlir::Location, mlir::OperationName, mlir::TypeRange, mlir::ValueRange, mlir::DictionaryAttr, mlir::BlockRange, unsigned int) (/home/ryuta/packages/flang/build/bin/../lib/libMLIRIR.so.13git+0x10e8d5)
 #5 0x00007fcec3033a30 mlir::Operation::create(mlir::OperationState const&) (/home/ryuta/packages/flang/build/bin/../lib/libMLIRIR.so.13git+0x10ea30)
 #6 0x00007fcec2fa9c80 mlir::OpBuilder::createOperation(mlir::OperationState const&) (/home/ryuta/packages/flang/build/bin/../lib/libMLIRIR.so.13git+0x84c80)
 #7 0x00007fcec4bd5bb6 mlir::CondBranchOp mlir::OpBuilder::create<mlir::CondBranchOp, mlir::Value&, mlir::Block*&, llvm::NoneType const&, mlir::Block*&, llvm::NoneType const&>(mlir::Location, mlir::Value&, mlir::Block*&, llvm::NoneType const&, mlir::Block*&, llvm::NoneType const&) Bridge.cpp:0:0
 #8 0x00007fcec4bd8557 (anonymous namespace)::FirConverter::genFIR(Fortran::parser::IfConstruct const&)::'lambda0'(mlir::Value)::operator()(mlir::Value) const Bridge.cpp:0:0
```

**Comment by vdonaldson** [comment]
Subroutine CMAUTO has an irreducible control flow graph.  Its CFG contains a strongly connected component with two distinct entry points, the statements with labels 200 and 220.  The IF construct generated from:

    IF (NLIN .LT. NLPP) GO TO 220

is rewritten to eliminate the goto, and the result is initially found to be structured (which at that point is correct).  So the goto target of the first IF statement, which is the call to BITPAT, is not marked as starting a new basic block.  Eventually the backward branch:

    IF (NLIN .GT. NLPP) GO TO 200

is processed, and at that point the first IF is marked as unstructured.  However, its goto target statement is never marked as starting a basic block.  The goto target of the first IF should be marked as starting a basic block when that IF is marked as unstructured.

**Comment by vdonaldson** [comment]
Fixed with PR #876

**Comment by oroppas** [comment]
I can confirm the fix. Thanks!
